src/tools/rbd_mirror/image_replayer/journal/Replayer.cc

```diff
@@ -405,14 +405,37 @@ bool Replayer<I>::notify_init_complete(std::unique_lock<ceph::mutex>& locker) {
 }
 
 template <typename I>
-void Replayer<I>::shut_down_local_journal_replay() {
+void Replayer<I>::wait_for_flush() {
   ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
 
+  // ensure that we don't have two concurrent local journal replay shut downs
+  dout(10) << dendl;
+  auto ctx = create_async_context_callback(
+    m_threads->work_queue, create_context_callback<
+      Replayer<I>, &Replayer<I>::handle_wait_for_flush>(this));
+  m_flush_tracker.wait_for_ops(ctx);
+}
+
+template <typename I>
+void Replayer<I>::handle_wait_for_flush(int r) {
+  dout(10) << "r=" << r << dendl;
+
+  shut_down_local_journal_replay();
```

Comment:
Just thinking. Isn't it theoretically possible that a new replay flush comes in the time window between the flush tracker's `wait_for_ops` completes and the shutdown started?

Reply:
I think you cannot start a flush if the state is not in `REPLAYING` mode -- and the first thing shut-down does it switch the state.

Reply:
indeed